test/jdk/java/nio/channels/DatagramChannel/SelectWhenRefused.java

```diff
@@ -94,9 +127,35 @@ public static void main(String[] args) throws IOException {
     }
 
     static void sendDatagram(DatagramChannel dc, SocketAddress remote)
-        throws IOException
-    {
-        ByteBuffer bb = ByteBuffer.wrap("Greetings!".getBytes());
+            throws IOException {
+        ByteBuffer bb = ByteBuffer.wrap(GREETINGS_MESSAGE.getBytes());
         dc.send(bb, remote);
     }
+
+    static boolean checkUnexpectedWakeup(Set<SelectionKey> selectedKeys) {
+        System.out.format("Received %d keys%n", selectedKeys.size());
+        for (SelectionKey key : selectedKeys) {
+            if (!key.isValid() || !key.isReadable()) {
+                System.out.println("Invalid or unreadable key: " + key);
+                continue;
+            }
+            try {
+                System.out.println("Attempting to read datagram from key: " + key);
+                DatagramChannel datagramChannel = (DatagramChannel) key.channel();
+                ByteBuffer buf = ByteBuffer.allocate(100);
+                SocketAddress sa = datagramChannel.receive(buf);
+                String message = new String(buf.array());
```

Comment:
You can't really use new String(buf.array()) here because you will read past the bytes that have been received. Also keep in mind that after receiving the position of the BB will be *at the end* of the message.

Reply:
Adding `buf.flip()` between 146 and 147 should be enough to address this right?

Reply:
No because buf.array() always return the same array of 100 bytes that the buffer wraps, regardless of the byte buffer position and limit.